pkg/controller/deployment/deployment_controller.go

```diff
@@ -446,23 +475,82 @@ func (dc *DeploymentController) handleErr(err error, key interface{}) {
 	dc.queue.Forget(key)
 }
 
-// claimReplicaSets uses NewReplicaSetControllerRefManager to classify ReplicaSets
-// and adopts them if their labels match the Deployment but are missing the reference.
-// It also removes the controllerRef for ReplicaSets, whose labels no longer matches
-// the deployment.
-func (dc *DeploymentController) claimReplicaSets(deployment *extensions.Deployment) error {
-	rsList, err := dc.rsLister.ReplicaSets(deployment.Namespace).List(labels.Everything())
+// getReplicaSetsForDeployment uses ControllerRefManager to reconcile
+// ControllerRef by adopting and orphaning.
+// It returns the list of ReplicaSets that this Deployment should manage.
+func (dc *DeploymentController) getReplicaSetsForDeployment(d *extensions.Deployment) ([]*extensions.ReplicaSet, error) {
+	rsList, err := dc.rsLister.ReplicaSets(d.Namespace).List(labels.Everything())
 	if err != nil {
-		return err
+		return nil, err
+	}
+	deploymentSelector, err := metav1.LabelSelectorAsSelector(d.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("deployment %s/%s has invalid label selector: %v", d.Namespace, d.Name, err)
+	}
+	cm := controller.NewReplicaSetControllerRefManager(dc.rsControl, d, deploymentSelector, controllerKind)
+	return cm.ClaimReplicaSets(rsList)
+}
+
+// getPodMapForReplicaSets scans the list of all Pods and returns a map from
+// RS UID to Pods controlled by that RS, based on the Pod's ControllerRef.
+func (dc *DeploymentController) getPodMapForReplicaSets(namespace string, rsList []*extensions.ReplicaSet) (map[types.UID]*v1.PodList, error) {
+	// List all Pods.
+	pods, err := dc.podLister.Pods(namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	// Group Pods by their controller (if it's in rsList).
+	podMap := make(map[types.UID]*v1.PodList, len(rsList))
+	for _, rs := range rsList {
+		podMap[rs.UID] = &v1.PodList{}
+	}
+	for _, pod := range pods {
+		// Ignore inactive Pods since that's what ReplicaSet does.
+		if !controller.IsPodActive(pod) {
+			continue
+		}
+		controllerRef := controller.GetControllerOf(pod)
+		if controllerRef == nil {
+			continue
+		}
+		// Only append if we care about this UID.
+		if podList, ok := podMap[controllerRef.UID]; ok {
+			podList.Items = append(podList.Items, *pod)
+		}
 	}
+	return podMap, nil
+}
 
-	deploymentSelector, err := metav1.LabelSelectorAsSelector(deployment.Spec.Selector)
+// getPodCountForReplicaSets scans the list of all Pods and returns the total
+// number of Pods controlled by all ReplicaSets in rsList, based on the Pod's
+// ControllerRef.
+func (dc *DeploymentController) getPodCountForReplicaSets(namespace string, rsList []*extensions.ReplicaSet) (int, error) {
```

Comment:
Seems like it has a lot of duplicate code. I would be in favor of dropping it and simply estimating the count in the call site (use `getPodMapForReplicaSets`)